### PR TITLE
Support for HTTP cookies

### PR DESCRIPTION
**Pull request by awaterme**
Hi Oliver,
I would like to propose an enhancement: Optional cookie management for HTTP sources (disabled by default).

Background: We have to integrate a proprietary WMS as source, which has difficulties if clients do not support HTTP session management through cookies. In other words: The client (mapproxy) has to accept and pass back session cookies.

I would be glad if you would agree to support such a scenario with MapProxy and accept this contribution.

So: Could you please review the changes? Important: I know almost nothing about python, so please let me know if something is not Ok. Luckily it is quite easy to make MapProxy support cookies. 
I can also offer to adjust the documentation. But before I start working on the docs, please let me know if you can image to accept this.

With best regards,
Andreas

### CODE REVIEW

File: mapproxy/client/http.py [hunk 1]
```diff
@@ -142,6 +145,9 @@ def __call__(self, ssl_ca_certs, url, username, password, insecure=False):
             handlers.append(authhandler)
             authhandler = urllib2.HTTPDigestAuthHandler(passman)
             handlers.append(authhandler)
+            if (manage_cookies):
```
Comment (olt) [inline mapproxy/client/http.py]: No need for parentheses in Python.

Comment (awaterme) [reply mapproxy/client/http.py]: Do you want me to remove them?

**Comment by olt** [comment]
Yes, why not. Just two questions to understand the reason behind this:

Does the server create a new session for each new request and the number of sessions is limited?
MapProxy will still start multiple sessions, depending on your deployment, but at least after restart of the server. Does that still work for you use-case?

File: mapproxy/client/http.py (hunk 1 above)
Comment (olt) [reply mapproxy/client/http.py]: Yes please. The other lines look good!

**Comment by awaterme** [comment]
Hi,
I suppose the server can handle this behaviour. It is currently used as backend for a web portal where it maintains the sessions for many users, without bigger problems for years now. So a couple of sessions for the client "MapProxy" will not cause problems. However we are not yet in production, so time will show.
I am quite happy if you consider accepting this and I am also happy with adjusting the parentheses :-)

**Comment by awaterme** [comment]
Shall I try to amend the documentation, too?

**Comment by awaterme** [comment]
Hi Oliver,

I have adjusted the docs now. I have also generated the docs; its seems Ok to me. Let me know otherwise.

Please check if my explanation regarding the cookie handling is correct: It is more as guess, because I don't know under which circumstances MapProxy will create new url_openers. I suppose per process...

I also added some links to #182 regarding basic authentication because I found the docs a bit misleading.

**Comment by awaterme** [comment]
Hi Oliver, are you still willing to accept this pull request?
I'd be rather glad...
Please let me know if anything is missing.
Best regards, Andreas

**Comment by m-kuhn** [comment]
What is the status of this? Could this be brought back into a mergeable state by rebasing on current master and fixing up the docs?

**Comment by weskamm** [comment]
closing due to inactivity

**Comment by awaterme** [comment]
Continued on https://github.com/mapproxy/mapproxy/pull/482

File: mapproxy/config_template/base_config/full_example.yaml [hunk 2]
```diff
@@ -425,10 +425,27 @@ sources:
     req:
       # username and password are extracted from the URL and do not show
       # up in log files
+      # ATTENTION: see https://github.com/mapproxy/mapproxy/issues/182
```
Comment (weskamm) [inline mapproxy/config_template/base_config/full_example.yaml:428]: please remove this line and move the information to the docs in the appropriate section

File: mapproxy/config_template/base_config/full_example.yaml [hunk 3]
```diff
@@ -425,10 +425,27 @@ sources:
...
 
+  # WMS source that requires authentication and session management
+  # through HTTP cookies
+  session_source:
+    type: wms
+    http:
+      # Accept session cookies and forward on subsequent requests
+      manage_cookies: true
+      # Use basic auth header directly
+      # see https://github.com/mapproxy/mapproxy/issues/182
```
Comment (weskamm) [inline mapproxy/config_template/base_config/full_example.yaml:441]: Same as above, please remove this line

**Pull request closed without merging**